### PR TITLE
Creating a product ad: WIP

### LINKED ISSUE #14: Create a Product ad

### PR DESCRIPTION
**Pull request by Slymee**

### CODE REVIEW

File: app/Http/Controllers/ProductController.php [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+                    'product_title' => $request->input('product_title'),
+                    'product_description' => $request->input('product_description'),
+                    'product_price' => $request->input('product_price'),
+                    'image_path' => $imagePath,
+                    'slug' => $request['slug'],
+                ]);
+                $productAd->categories()->attach($request->input('parent_category'));
+                $productAd->categories()->attach($request->input('sub_category'));
+                $productAd->categories()->attach($request->input('sub_sub_category'));
+                $tagNames = $request->input('product_tags', []);
```
Comment (aashish-t-magar) [inline app/Http/Controllers/ProductController.php]: sub sub category yo matra assign garney

File: app/Http/Controllers/ProductController.php [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+                $productAd->tags()->saveMany(
+                    array_map(function ($tagName) {
+                        return new Tag(['tag_name' => $tagName]);
+                    }, $tagNames)
+                );
```
Comment (aashish-t-magar) [inline app/Http/Controllers/ProductController.php]: attach or sync

File: app/Models/Tag.php [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    use HasFactory;
+    protected $table = 'tags';
+    protected $fillable = [
+        'product_id',
+        'tag_name',
+    ];
+
+    public function productAds(){
+        return $this->belongsTo(Product::class, 'product_ads');
+    }
```
Comment (aashish-t-magar) [inline app/Models/Tag.php]: relation ship ko name milena......psr -12

File: app/Models/Product.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    protected $table='products';
+
+    protected $fillable=[
+        'user_id',
+        'product_title',
+        'product_description',
+        'product_price',
+        'image_path',
+        'slug'
+    ];
```
Comment (aashish-t-magar) [inline app/Models/Product.php]: add category_id...one category one product

if want to associate many...use tags

File: app/Models/CategoryProduct.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+namespace App\Models;
+
+use Illuminate\Database\Eloquent\Factories\HasFactory;
+use Illuminate\Database\Eloquent\Model;
+
+class CategoryProduct extends Model
+{
+    use HasFactory;
+
+    protected $table = 'category_product';
```
Comment (aashish-t-magar) [inline app/Models/CategoryProduct.php]: psr-12

File: app/Http/Controllers/ProductController.php [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+     * Remove the specified resource from storage.
+     */
+    public function destroy(string $productId): \Illuminate\Http\RedirectResponse
+    {
+        try {
+            Product::find($productId)->delete();
+            return redirect()->back()->with('message', 'Product Deleted');
+        }catch (\Exception $e){
+            return redirect()->back()->with('message', $e->getMessage());
+        }
```
Comment (aashish-t-magar) [inline app/Http/Controllers/ProductController.php]: delete image also

File: app/Http/Requests/CreateProductAdRequest.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
+<?php
+
+namespace App\Http\Requests;
+
+use Illuminate\Foundation\Http\FormRequest;
+
+class CreateProductAdRequest extends FormRequest
+{
+    /**
+     * Determine if the user is authorized to make this request.
```
Comment (aashish-t-magar) [inline app/Http/Requests/CreateProductAdRequest.php]: create or update  use same

File: app/Http/Requests/ProductEditUpdateRequest.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+     * Determine if the user is authorized to make this request.
+     */
+    public function authorize(): bool
+    {
+        return true;
+    }
+
+    /**
+     * Get the validation rules that apply to the request.
+     *
```
Comment (aashish-t-magar) [inline app/Http/Requests/ProductEditUpdateRequest.php]: remove

File: app/Models/Product.php [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    public function categories(): \Illuminate\Database\Eloquent\Relations\BelongsToMany
+    {
+        return $this->belongsToMany(Category::class, 'product_categories');
+    }
+
+    public function tags(): \Illuminate\Database\Eloquent\Relations\HasMany
+    {
+        return $this->hasMany(Tag::class, 'product_id');
+    }
+}
```
Comment (aashish-t-magar) [inline app/Models/Product.php]: use another relatinship

File: app/Http/Requests/CreateProductRequest.php [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+     * @return array<string, \Illuminate\Contracts\Validation\ValidationRule|array<mixed>|string>
+     */
+    public function rules(): array
+    {
+        $rules = [
+            'user_id' => ['required', 'bail', 'exists:users,id'],
+            'product_title' => ['required', 'bail'],
+            'product_description' => ['required', 'bail'],
+            'product_price' => ['required', 'bail'],
+            'product_image' => ['required', 'image', 'mimes:jpeg,png,jpg,gif,svg', 'max:2048'],
```
Comment (A4family) [inline app/Http/Requests/CreateProductRequest.php:29]: sometimes rule

**Review by aashish-t-magar — APPROVED** [review]

**Pull request merged**